test/extended/router/gatewayapicontroller.go

```diff
@@ -0,0 +1,507 @@
+		g.By("Getting the default domain")
+		defaultIngressDomain, err := getDefaultIngressClusterDomainName(oc, time.Minute)
+		o.Expect(err).NotTo(o.HaveOccurred(), "failed to find default domain name")
+		defaultDomain := strings.Split(defaultIngressDomain, "apps.")[1]
+
+		g.By("Create the default API Gateway")
+		gw := names.SimpleNameGenerator.GenerateName("gateway-")
+		gateways = append(gateways, gw)
+		_, gwerr := createAndCheckGateway(oc, gw, gatewayClassName, defaultDomain)
+		o.Expect(gwerr).NotTo(o.HaveOccurred(), "failed to create Gateway")
+
+		g.By("Verify the gateway's LoadBalancer service and DNSRecords")
+		// check LB service
+		lbExternalIP, lberr := oc.AdminKubeClient().CoreV1().Services("openshift-ingress").Get(context.Background(), gw+"-openshift-default", metav1.GetOptions{})
+		e2e.Logf("The load balancer external IP is: %v", lbExternalIP.Status.LoadBalancer.Ingress[0].Hostname)
+		o.Expect(lberr).NotTo(o.HaveOccurred())
+
+		gwapiClient := gatewayapiclientset.NewForConfigOrDie(oc.AdminConfig())
+		gwlist, haerr := gwapiClient.GatewayV1().Gateways("openshift-ingress").Get(context.Background(), gw, metav1.GetOptions{})
+		e2e.Logf("The gateway hostname address is %v ", gwlist.Status.Addresses[0].Value)
+		o.Expect(haerr).NotTo(o.HaveOccurred())
+		o.Expect(lbExternalIP.Status.LoadBalancer.Ingress[0].Hostname).To(o.Equal(gwlist.Status.Addresses[0].Value))
+
+		// get the dnsrecord name
+		dnsRecordName, err := oc.AsAdmin().WithoutNamespace().Run("get").Args("-n", "openshift-ingress", "dnsrecord", "-l", "gateway.networking.k8s.io/gateway-name="+gw, "-o=jsonpath={.items[0].metadata.name}").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		e2e.Logf("The gateway API dnsrecord name is: %v", dnsRecordName)
+		// check whether status of published dns record of the gateway is True
+		dnsRecordstatus, err := oc.AsAdmin().WithoutNamespace().Run("get").Args("-n", "openshift-ingress", "dnsrecord", dnsRecordName, `-o=jsonpath={.status.zones[0].conditions[0].status}`).Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(dnsRecordstatus).To(o.Equal("True"))
+	})
+
+	g.It("Ensure HTTPRoute object is created", func() {
+		g.By("Getting the default domain")
+		defaultIngressDomain, err := getDefaultIngressClusterDomainName(oc, time.Minute)
+		o.Expect(err).NotTo(o.HaveOccurred(), "failed to find default domain name")
+		defaultDomain := strings.Split(defaultIngressDomain, "apps.")[1]
+
+		g.By("Create a custom Gateway for the HTTPRoute")
+		gw := names.SimpleNameGenerator.GenerateName("gateway-")
+		gateways = append(gateways, gw)
+		_, gwerr := createAndCheckGateway(oc, gw, gatewayClassName, defaultDomain)
+		o.Expect(gwerr).NotTo(o.HaveOccurred(), "failed to create Gateway")
+
+		g.By("Create the http route using the custom gateway")
+		httproutes = append(httproutes, "test-httproute")
+		defaultRoutename := "test-hostname.gwapi." + defaultDomain
+		createHttpRoute(oc, gw, "test-httproute", defaultRoutename, "echo-pod-"+gw)
```

Comment:
the order is wrong
createHttpRoute(oc, "test-httproute", defaultRoutename, gw, "echo-pod-"+gw)